File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Views/Components/FiftyPercentOffZoneView.swift
import SwiftUI

struct FiftyPercentOffZoneView: View {
    private var wishlistManager: WishlistManager { WishlistManager.shared }

    // Props from SDUI JSON
    var title: String = "50% OFF ZONE"
    var subtitle: String = "Half the price, double the joy!"
    var bannerImage: String =
        "https://png.pngtree.com/png-vector/20240125/ourmid/pngtree-grocery-shopping-bag-isolated-png-image_11549419.png"
    var discountText: String = "50%"
    var categoryId: String?
    var subCategoryIds: [String] = []
    var isGrocery: Bool = false

    @State private var products: [Product] = []
    @State private var isLoading = true
    @State private var hasLoaded = false
    @State private var fetchTask: Task<Void, Never>? = nil

    var body: some View {
        VStack(spacing: 0) {
            // Header Banner
            ZStack(alignment: .leading) {
                LinearGradient(
                    gradient: Gradient(colors: [Color(hex: "#F0F9FF"), Color(hex: "#E0F2FE")]),
                    startPoint: .leading,
                    endPoint: .trailing
                )

                HStack {
                    VStack(alignment: .leading, spacing: 4) {
                        HStack(alignment: .bottom, spacing: 4) {
                            Text(discountText)
                                .font(.system(size: 42, weight: .black))
                                .italic()
                                .foregroundColor(Color(hex: "#2563EB"))

                            VStack(alignment: .leading, spacing: 0) {
                                Text("OFF")
                                    .font(.system(size: 14, weight: .heavy))
                                    .italic()
                                    .foregroundColor(Color(hex: "#2563EB"))
                                Text("ZONE")
                                    .font(.system(size: 14, weight: .heavy))
                                    .italic()
                                    .foregroundColor(Color(hex: "#3B82F6"))
                            }
                            .padding(.bottom, 6)

                            Image(systemName: "sparkles")
                                .foregroundColor(Color(hex: "#3B82F6"))
                                .padding(.bottom, 12)
                        }

                        Text(subtitle)
                            .font(.system(size: 13, weight: .medium))
                            .foregroundColor(Color(hex: "#1F2937"))
                    }
                    .padding(.leading, 20)

                    Spacer()

                    AsyncImage(url: URL(string: bannerImage)) { image in
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                    } placeholder: {
                        Color.clear
                    }
                    .frame(width: 140, height: 100)
                    .padding(.trailing, 10)
                }
            }
            .frame(height: 140)

            // Product List
            if isLoading && !hasLoaded {
                HStack {
                    Spacer()
                    ProgressView()
                    Spacer()
                }
                .padding(20)
            } else if products.isEmpty {
                VStack(spacing: 8) {
                    Image(systemName: "tag.slash")
                        .font(.system(size: 28))
                        .foregroundColor(.gray)
                    Text("No deals available right now")
                        .font(.system(size: 13))
                        .foregroundColor(.gray)
                }
                .frame(maxWidth: .infinity)
                .padding(20)
            } else {
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 12) {
                        ForEach(products) { product in
                            NavigationLink(
                                destination: ProductDetailView(
                                    productId: product.id,
                                    productFragment: product
                                )
                            ) {
                                ProductCardWithWishlist(
                                    product: product,
                                    width: 150
                                )
                            }
                            .buttonStyle(PlainButtonStyle())
                        }
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 16)
                }
            }

            // See All Button
            NavigationLink(
                destination: CategoryProductsListView(
                    categoryId: categoryId,
                    subCategoryIds: subCategoryIds,
                    isGrocery: isGrocery,
                    title: title,
                    minimumDiscount: 50
                )
            ) {
                HStack {
                    Text("See all")
                        .font(.system(size: 14, weight: .semibold))
                    Image(systemName: "chevron.right")
                        .font(.system(size: 12, weight: .bold))
                }
                .foregroundColor(Color(hex: "#4F46E5"))
                .frame(maxWidth: .infinity)
                .padding(.vertical, 12)
                .background(Color(hex: "#EEF2FF"))
                .cornerRadius(8)
                .padding(.horizontal, 16)
                .padding(.bottom, 16)
            }
        }
        .background(Color.white)
        .padding(.vertical, 12)
        .task {
            guard !hasLoaded else { return }
            // Cancel any existing fetch to avoid duplicates
            fetchTask?.cancel()
            let task = Task {
                await loadProducts()
            }
            fetchTask = task
            await task.value
        }
    }

    private func loadProducts() async {
        // Prevent re-entry if already loaded or currently loading
        guard !hasLoaded else { return }

        isLoading = true

        do {
            // Check for cancellation before starting network request
            try Task.checkCancellation()

            var fetchedProducts: [Product] = []

            if !subCategoryIds.isEmpty {
                // Fetch by sub-category IDs (most specific)
                fetchedProducts = try await APIService.shared.fetchProductsBySubCategoryIds(
                    subCategoryIds, limit: 30)
            } else if let catId = categoryId, !catId.isEmpty {
                // Fetch by category ID
                fetchedProducts = try await APIService.shared.fetchProducts(
                    limit: 30, categoryId: catId)
            } else {
                // Fallback: fetch general products
                fetchedProducts = try await APIService.shared.fetchProducts(limit: 30)
            }

            // Check cancellation after network call
            try Task.checkCancellation()

            // Filter to only products with >=50% discount
            let discounted = fetchedProducts.filter { product in
                if let discount = product.discountPercentage, discount >= 50 {
                    return true
                }
                // Calculate from MRP if discountPercentage not set
                if let mrp = product.mrp, mrp > product.price {
                    let calculatedDiscount = Int(((mrp - product.price) / mrp) * 100)
                    return calculatedDiscount >= 50
                }
                return false
            }

            // Show discounted products, or first 10 of fetched if no discounts found
            self.products =
                discounted.isEmpty
                ? Array(fetchedProducts.prefix(10))
                : Array(discounted.prefix(10))

            self.hasLoaded = true
            self.isLoading = false
        } catch is CancellationError {
            // SwiftUI cancelled the task — do NOT retry, just exit silently
            AppLogger.debug("[FiftyPercentOffZone] Task cancelled, not retrying")
            return
        } catch let error as NSError
            where error.domain == NSURLErrorDomain && error.code == -999
        {
            // URLSession cancelled (error -999) — same as CancellationError, exit silently
            AppLogger.debug("[FiftyPercentOffZone] URLSession cancelled (-999), not retrying")
            return
        } catch {
            AppLogger.debug("[FiftyPercentOffZone] Error fetching products: \(error)")
            // Mark as loaded even on error to prevent infinite retry loop
            self.hasLoaded = true
            self.isLoading = false
        }
    }
}

// MARK: - Product Card with Working Wishlist

private struct ProductCardWithWishlist: View {
    let product: Product
    let width: CGFloat
    private var wishlistManager: WishlistManager { WishlistManager.shared }

    var isWishlisted: Bool {
        wishlistManager.isInWishlist(productId: product.id)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Image Container
            ZStack(alignment: .topTrailing) {
                Color(hex: "#F3F4F6")

                if let imageUrl = product.images.first, let url = URL(string: imageUrl) {
                    CachedAsyncImage(url: url) { image in
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fill)
                    } placeholder: {
                        Color(hex: "#F3F4F6")
                    }
                    .frame(width: width, height: 160)
                    .clipped()
                } else {
                    Image(systemName: "photo")
                        .font(.system(size: 40))
                        .foregroundColor(.gray)
                        .frame(width: width, height: 160)
                }

                // Wishlist Button (functional)
                Button(action: {
                    HapticManager.shared.impact(style: .medium)
                    Task {
                        _ = await wishlistManager.toggleWishlist(productId: product.id)
                    }
                }) {
                    Circle()
                        .fill(Color.white)
                        .frame(width: 32, height: 32)
                        .shadow(color: .black.opacity(0.1), radius: 2, x: 0, y: 1)
                        .overlay(
                            Image(systemName: isWishlisted ? "heart.fill" : "heart")
                                .font(.system(size: 16))
                                .foregroundColor(
                                    isWishlisted
                                        ? Color(hex: "#EF4444") : Color(hex: "#4F46E5"))
                        )
                }
                .padding(8)

                // Discount Badge
                if let discount = product.discountPercentage, discount > 0 {
                    Text("-\(discount)%")
                        .font(.system(size: 10, weight: .bold))
                        .foregroundColor(.white)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Color(hex: "#EF4444"))
                        .cornerRadius(4)
                        .padding(8)
                        .frame(maxWidth: .infinity, alignment: .topLeading)
                }
            }
            .frame(height: 160)

            // Content
            VStack(alignment: .leading, spacing: 4) {
                Text(product.name)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(Color(hex: "#111827"))
                    .lineLimit(2)
                    .frame(height: 40, alignment: .topLeading)

                HStack(alignment: .center) {
                    VStack(alignment: .leading, spacing: 0) {
                        Text("₹\(Int(product.price))")
                            .font(.system(size: 16, weight: .bold))
                            .foregroundColor(Color(hex: "#4F46E5"))

                        if let mrp = product.mrp, mrp > product.price {
                            Text("₹\(Int(mrp))")
                                .font(.system(size: 10))
                                .foregroundColor(Color(hex: "#9CA3AF"))
                                .strikethrough()
                        }
                    }

                    Spacer()

                    // Rating
                    if let rating = product.rating {
                        HStack(spacing: 2) {
                            Image(systemName: "star.fill")
                                .font(.system(size: 10))
                                .foregroundColor(Color(hex: "#F59E0B"))
                            Text(String(format: "%.1f", rating))
                                .font(.system(size: 12, weight: .semibold))
                                .foregroundColor(Color(hex: "#B45309"))
                        }
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Color(hex: "#FFFBEB"))
                        .cornerRadius(4)
                    }
                }
                .padding(.top, 4)
            }
            .padding(12)
        }
        .frame(width: width)
        .background(Color.white)
        .cornerRadius(16)
        .shadow(color: .black.opacity(0.05), radius: 4, x: 0, y: 2)
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/ui/components/FiftyPercentOffZoneView.kt
package com.ecommerceearn.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowRight
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.ecommerceearn.app.data.model.Product
import com.ecommerceearn.app.data.remote.NetworkClient
import com.ecommerceearn.app.ui.components.ProductCardView

@Composable
fun FiftyPercentOffZoneView(
    title: String = "50% OFF ZONE",
    subtitle: String = "Half the price, double the joy!",
    bannerImage: String = "https://png.pngtree.com/png-vector/20240125/ourmid/pngtree-grocery-shopping-bag-isolated-png-image_11549419.png",
    discountText: String = "50%"
) {
    var products by remember { mutableStateOf<List<Product>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }

    LaunchedEffect(Unit) {
        try {
            val fetched = NetworkClient.apiService.getProductsRaw(10).products
            products = fetched
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            isLoading = false
        }
    }

    Column(
        modifier = Modifier
            .padding(vertical = 12.dp)
            .background(Color.White)
    ) {
        // Header Banner
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(140.dp)
                .background(
                    Brush.horizontalGradient(
                        colors = listOf(safeParseColor("#F0F9FF"), safeParseColor("#E0F2FE"))
                    )
                )
        ) {
            Row(
                modifier = Modifier.fillMaxSize(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Text Content
                Column(
                    modifier = Modifier
                        .padding(start = 20.dp)
                        .weight(1f)
                ) {
                    Row(verticalAlignment = Alignment.Bottom) {
                         Text(
                            text = discountText,
                            fontSize = 42.sp,
                            fontWeight = FontWeight.Black,
                            fontStyle = FontStyle.Italic,
                            color = safeParseColor("#2563EB")
                        )
                        Spacer(modifier = Modifier.width(4.dp))
                        Column(modifier = Modifier.padding(bottom = 6.dp)) {
                            Text(
                                text = "OFF",
                                fontSize = 14.sp,
                                fontWeight = FontWeight.ExtraBold,
                                fontStyle = FontStyle.Italic,
                                color = safeParseColor("#2563EB")
                            )
                            Text(
                                text = "ZONE",
                                fontSize = 14.sp,
                                fontWeight = FontWeight.ExtraBold,
                                fontStyle = FontStyle.Italic,
                                color = safeParseColor("#3B82F6")
                            )
                        }
                    }
                    Text(
                        text = subtitle,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.Medium,
                        color = safeParseColor("#1F2937")
                    )
                }

                // Image
                AsyncImage(
                    model = bannerImage,
                    contentDescription = null,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .width(140.dp)
                        .height(100.dp)
                        .padding(end = 10.dp)
                )
            }
        }

        // Product List
        if (isLoading) {
            Box(modifier = Modifier.fillMaxWidth().height(200.dp), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
        } else if (products.isNotEmpty()) {
            LazyRow(
                contentPadding = PaddingValues(16.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(products) { product ->
                    ProductCardView(product = product, modifier = Modifier.width(150.dp))
                }
            }
        }

        // See All Button
        Box(
            modifier = Modifier
                .padding(horizontal = 16.dp)
                .padding(bottom = 16.dp)
                .fillMaxWidth()
                .clip(RoundedCornerShape(8.dp))
                .background(safeParseColor("#EEF2FF"))
                .clickable { }
                .padding(vertical = 12.dp),
            contentAlignment = Alignment.Center
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = "See all",
                    fontSize = 14.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = safeParseColor("#4F46E5")
                )
                Icon(
                    imageVector = Icons.Default.KeyboardArrowRight,
                    contentDescription = null,
                    tint = safeParseColor("#4F46E5"),
                    modifier = Modifier.size(16.dp)
                )
            }
        }
    }
}
